history-service/src/storage/attempt_store.ts

```diff
@@ -145,6 +146,58 @@ class AttemptStore implements IAttemptStore {
       totalCount,
     };
   }
+
+  async removeHistoryOwner(userId: number): Promise<void> {
+    // Delete HistoryOwner and retrieve attempts
+    const resultAttemptId: { attempt_id: number; }[] = (await this.dbConn
+      .getDataSource()
+      .createQueryBuilder()
+      .delete()
+      .from(HistoryOwnerEntity)
+      .where('user_id = :userId', { userId })
+      .returning('attempt_id')
+      .execute()
+    ).raw;
+    if (!resultAttemptId) {
+      return;
+    }
+    const deletedUserAttempts = resultAttemptId.map(
+      (res: { attempt_id: number; }) => res.attempt_id,
+    );
+    if (!deletedUserAttempts.length) {
+      return;
+    }
+
+    // Retrieve deleted owner's attempts that has other owners
+    const otherUserAttempts: HistoryAttemptEntity[] = await this.dbConn
+      .getHistoryRepo()
+      .createQueryBuilder('histories')
+      .innerJoinAndSelect('histories.users', 'history_owners')
+      .where('histories.attempt_id IN (:...listOfIds)', { listOfIds: deletedUserAttempts })
+      .getMany();
+    const attemptsWithOtherUser = new Set();
+    otherUserAttempts.map(
+      (res: HistoryAttemptEntity) => attemptsWithOtherUser.add(res.attemptId),
+    );
```

Comment:
You can actually skip this duplicate checking step, the PK constraint on the column ensures that there wont be duplicates.

Reply:
This `set` isn't used for duplicate checking but is used for line 184 to verify if the attempt still has another user in the db. In a sense, this set consist of attempts of the deleted user that are still to tagged to another user. Am using the set for faster checking.

Reply:
Sorry, this the confusion was in inner join not returning entries with 0 users. This implementation is fine then.